staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -62,11 +62,13 @@ type Validator struct {
 // NewValidator returns compiles all the CEL programs defined in x-kubernetes-validations extensions
 // of the Structural schema and returns a custom resource validator that contains nested
 // validators for all items, properties and additionalProperties that transitively contain validator rules.
-// Returns nil only if there no validator rules in the Structural schema. May return a validator containing
-// only errors.
+// Returns nil if there are no validator rules in the Structural schema. May return a validator containing only errors.
 // Adding perCallLimit as input arg for testing purpose only. Callers should always use const PerCallLimit as input
 func NewValidator(s *schema.Structural, isResourceRoot bool, perCallLimit uint64) *Validator {
-	return validator(s, true, model.SchemaDeclType(s, isResourceRoot), perCallLimit)
+	if !hasXValidations(s) {
```

Comment:
I'm not sure where this is called from, but maybe only run this check at the root to avoid recursively checking at multiple levels?
```suggestion
	if isResourceRoot && !hasXValidations(s) {
```

Reply:
This is called by the defaulter where isResourceRoot will be false and we want to skip if there are no CEL rules.

This call is also much simpler, it only has 2 non-test callsites:  1 for defaulters when the init the CEL rule, run it and then throw it away, and 1 for when a CRD is loaded for storage

Reply:
ok, sounds good